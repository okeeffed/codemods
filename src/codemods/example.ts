import prettier from "prettier";
import { Project, SyntaxKind } from "ts-morph";

export const example = async (filePath: string, project: Project) => {
  // Add the source file you want to transform
  const sourceFile = project.addSourceFileAtPath(filePath);

  // Find all function declarations in the file
  sourceFile.getFunctions().forEach((func) => {
    // Get the name of the function
    const functionName = func.getName();

    if (!functionName) return;

    // We need to extract the return statement's content
    // Assuming the function body is simple and contains a single return statement
    const returnStatement = func
      .getBody()
      // @ts-ignore
      ?.getStatements()
      .find((s: any) => s.getKind() === SyntaxKind.ReturnStatement);

    // Extracting the text to return directly
    let returnText = returnStatement?.getChildAtIndex(1)?.getText() || "{}"; // Default to '{}' in case of no return

    // Create the new arrow function expression
    const arrowFunction = `const ${functionName} = () => (${returnText});`;

    // Replace the old function declaration with the new arrow function
    func.replaceWithText(arrowFunction);
  });

  // Save the changes to the file
  const text = sourceFile.getText();

  const prettierConfig = prettier.resolveConfigFile();

  // Format the result using Prettier
  const formatted = prettier.format(text, {
    ...prettierConfig,
    parser: "typescript",
  });

  return formatted;
};
