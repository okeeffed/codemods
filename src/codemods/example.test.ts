import { Project } from "ts-morph";
import { example } from "./example";

const EXAMPLE_TS = `function mockExample() {
	return {
			a: true,
	};
}`;

describe("example codemod", () => {
  test("expect snapshot to match", async () => {
    // Initialize a new Project
    const project = new Project({
      useInMemoryFileSystem: true,
    });

    // Add a file directly in the in-memory file system
    const filePath = "/mock/example.ts";

    project.createSourceFile(filePath, EXAMPLE_TS, { overwrite: true });
    const result = await example(filePath, project);

    expect(result).toMatchSnapshot();
  });
});
