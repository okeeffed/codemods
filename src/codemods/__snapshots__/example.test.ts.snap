// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example codemod expect snapshot to match 1`] = `
"const mockExample = () => ({
  a: true,
});
"
`;
